### PR TITLE
Nimble RPX Port

### PR DESCRIPTION
**Pull request by Fangal-Airbag**
This ports the Nimble patcher to wut and allows for rpx building. It does not include the Inkay patches as those seem to break a lot of the times for unknown reasons. The code also already allows for the creation of a channel/forwarder, as well as being fully aroma ready too!

Enjoy!

### CODE REVIEW

**Comment by Fangal-Airbag** [comment]
Welp at the current moment the Patcher is now fully functional!

File: .gitignore [hunk 1]
```diff
@@ -1,9 +0,0 @@
-build_files/
-build/
-
-arm_*.elf
-arm_*.bin
-arm_*_bin.h
-*.map
-
-iosu_payload/
```
Comment (QuarkTheAwesome) [inline .gitignore:9]: Would be good to keep a gitignore, if there are conflicts you can add or remove them as you need

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +0,0 @@
-FROM quarktheawesome/devkitpro-arm-ppc:r56_r39
-
-RUN apt-get update && apt-get install -y --no-install-recommends \
-  xxd \
-  && rm -rf /var/lib/apt/lists/*
-
-WORKDIR /app
-CMD make -j$(nproc)
```
Comment (QuarkTheAwesome) [inline Dockerfile:8]: Should keep the Dockerfile too, it helps a lot for people who don't have the perfect setup

File: Makefile.Aroma [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+DEPENDS	:=	$(OFILES:.o=.d)
+
+#-------------------------------------------------------------------------------
+# main targets
+#-------------------------------------------------------------------------------
+all	:	$(OUTPUT).wuhb
+
+$(OUTPUT).wuhb	:	$(OUTPUT).rpx
+$(OUTPUT).rpx	:	$(OUTPUT).elf
```
Comment (QuarkTheAwesome) [inline Makefile.Aroma:169]: I thiiink the Aroma rpx should be compatible with non-Aroma systems too, thus not requiring a seperate makefile? Not 100% sure on that, worth testing though

File: README.md [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
-// OpenSSL 1.0.0f 4 Jan 2012
-// @ E1019F50
-int ssl_verify_cert_chain(SSL *s,STACK_OF(X509) *sk);
+To create the channel version follow the same steps above to get `nimble_patcher.rpx`. Once done go to the folder named `code` inside the channel directory and place your rpx inside. Download NUSPacker.jar from [here](https://bitbucket.org/timogus/nuspacker/src/master/) and place it next to `channel` and create a folder named `WUP` in that same directory, afterwards run the following command in a terminal or cmd window (Replace the zeros with the Wii U common key)
+
+```shell
+java -jar NUSPacker.jar -in channel -out WUP -encryptKeyWith 00000000000000000000000000000000
 ```
+
+Afterwards you may place the contents from the `WUP` folder into the `install` folder on your SD card and then install via WUP Installer.
```
Comment (QuarkTheAwesome) [inline README.md:21]: I don't mean to be rough on the people who like channels, but, like, are we sure we want these instructions here? It sounds like not a great thing to do for Nimble

File: README.md [hunk 5]
```diff
@@ -1,16 +1,25 @@
-# NoSSL patch for the WiiU FW version 5.5.5
+# RPX port of Nimble
```
Comment (QuarkTheAwesome) [inline README.md:1]: We can probably just consider this to be the canonical version of Nimble, unless people really miss the title_patcher stuff.

File: channel/code/app.xml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+<app type="complex" access="777">
+  <version type="unsignedInt" length="4">1</version>
+  <os_version type="hexBinary" length="8">000500101000400A</os_version>
+  <title_id type="hexBinary" length="8">0005000014946100</title_id>
+  <title_version type="hexBinary" length="2">0000</title_version>
+  <sdk_version type="unsignedInt" length="4">20811</sdk_version>
+  <app_type type="hexBinary" length="4">80000000</app_type>
+  <group_id type="hexBinary" length="4">000004691</group_id>
+  <os_mask type="hexBinary" length="32">0000000000000000000000000000000000000000000000000000000000000000</os_mask>
+</app>
```
Comment (QuarkTheAwesome) [inline channel/code/app.xml:11]: Maybe all the channel stuff can be moved to the meta folder alonside the aroma/hbl meta?

File: source/main.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        titleID == MII_MAKER_EUR_TITLE_ID)
+    {
+        isChannel = false;
+    }
+
+    if (isChannel)
+    {
+        WHBLogPrintf("Is channel");
+        AXInit(); 
+        // ^ I hate this, it's just an issue with the provided bootSound.btsnd in this repo. You can use another one and then repack via NUSPacker if you want.
```
Comment (QuarkTheAwesome) [inline source/main.cpp:60]: I hate this too! Is there a silent bootSound we can use instead, maybe?

File: source/main.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+    bool isChannel = true;
+
+    if (titleID == HBL_TITLE_ID ||
+        titleID == MII_MAKER_JPN_TITLE_ID ||
+        titleID == MII_MAKER_USA_TITLE_ID ||
+        titleID == MII_MAKER_EUR_TITLE_ID)
+    {
+        isChannel = false;
+    }
```
Comment (QuarkTheAwesome) [inline source/main.cpp:54]: Is this code duplicated from Proc.cpp?

File: source/utils/dirent.h [hunk 9]
```diff
@@ -0,0 +1,1166 @@
+/*
+ * Dirent interface for Microsoft Visual Studio
```
Comment (QuarkTheAwesome) [inline source/utils/dirent.h:2]: Why is this here?

**Review by QuarkTheAwesome — CHANGES_REQUESTED** [review]
Overall quite good, just some minor things that could use some love. Should be a great option for people who aren't running Tiramisu or want to try Pretendo just once.

File: Makefile.Aroma (hunk 3 above)
Comment (Fangal-Airbag) [reply Makefile.Aroma:169]: That I rly wasn’t sure of but I’ll test

File: README.md (hunk 4 above)
Comment (Fangal-Airbag) [reply README.md:21]: K will remove lol

File: README.md (hunk 5 above)
Comment (Fangal-Airbag) [reply README.md:1]: I was waiting for it to the PR to actually be pushed to the repo but oki

File: channel/code/app.xml (hunk 6 above)
Comment (Fangal-Airbag) [reply channel/code/app.xml:11]: Yeah will do I dunno why I didn’t

File: source/main.cpp (hunk 7 above)
Comment (Fangal-Airbag) [reply source/main.cpp:60]: I’ll look around, if not I can just delete it or the silent bootSound idea

File: source/main.cpp (hunk 8 above)
Comment (Fangal-Airbag) [reply source/main.cpp:54]: I don’t wanna answer that…

_yes_

File: source/utils/dirent.h (hunk 9 above)
Comment (Fangal-Airbag) [reply source/utils/dirent.h:2]: It was in ultils before I’ll just delete it

File: Makefile.Aroma (hunk 3 above)
Comment (Fangal-Airbag) [reply Makefile.Aroma:169]: Yup they seem to be, I'll get rid of the separate makefile

File: Dockerfile (hunk 2 above)
Comment (Fangal-Airbag) [reply Dockerfile:8]: Imma need some help on that

**Review by QuarkTheAwesome — APPROVED** [review]
Looks good, I'll talk to you about Docker when I'm on a machine that can run it.
Will merge once I can test back home, couple days tops.

**Comment by Fangal-Airbag** [comment]
Welp added Docker now!

**Review by QuarkTheAwesome — APPROVED** [review]

**Comment by oscie57** [comment]
...is this gonna get merged or what lmao, only been half a year

**Comment by TermOfficial** [comment]
Is there a way you could somehow have the NoSSL patches only w/o redirecting to pretendo servers?

**Comment by QuarkTheAwesome** [comment]
Sorry about how this turned out. Nimble is focusing on Tiramisu going forward, any Aroma compatibility will likely end up in Inkay.
